### PR TITLE
Fix loss calculation in TFXXXForTokenClassification models

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

The current loss calculation in `TFFunnelForTokenClassification` (and other TF token classification models) don't align with the loss in `FunnelForTokenClassification`

TF
https://github.com/huggingface/transformers/blob/6ac77534bfe97c00e0127bb4fc846ae0faf1c9c5/src/transformers/models/funnel/modeling_tf_funnel.py#L1709

PT
https://github.com/huggingface/transformers/blob/6ac77534bfe97c00e0127bb4fc846ae0faf1c9c5/src/transformers/models/funnel/modeling_funnel.py#L1470-L1481

(which further uses the `attention_mask`).

This PR aims to fix this. Currently only `TFFunnelForTokenClassification` is fixed. I would like to have some feedback first before fixing all involved models.

## More information 1

Check the loss difference between PT/TF version 300 times,

For `(TF)BertForTokenClassification`
```
without this PR:

  max_diff:   0.022978603839874268
  mean_diff: 0.006492746472358704

with this PR (applied locally):

  max_diff:   1.7881393432617188e-07
  mean_diff: 3.3974647521972653e-08
```

For `(TF)FunnelForTokenClassification`
```
without this PR:

  max_diff:   0.2821081280708313
  mean_diff: 0.05486685564120611

with this PR:

  max_diff:   3.5762786865234375e-07
  mean_diff: 6.020069122314453e-08
```

## More information 2

Current transformers's version doesn't have a PT-TF test that tests the equivalence in the cases where `labels` is passed to the models. In PR #15256, such a test is introduced (to make sure more PT/TF equivalence). The inconsistency of loss calculation in PT/TF `XXXForTokenClassification` models requires the error tolerance to be not smaller than `0.2` in order to pass the test (and this doesn't even guarantee).

https://github.com/huggingface/transformers/blob/6134cc69527aa9df9a4f2bc9e545222456a34524/tests/test_modeling_tf_common.py#L413-L441

It would be better to address this inconsistency, and lower that threshold once this PR is merged.

### CODE REVIEW

**Review by sgugger — COMMENTED** [review]
I think in this instance it's the PyTorch model that is incorrect, and it should use the -100s in the labels instead of the attention mask to determine the logits/labels to ignore.

**Comment by LysandreJik** [comment]
This is a breaking change in case the attention mask does not align with the `-100` labels, but I think that's okay.

The token classification models are the only models that have this feature, which (AFAICT) isn't mentioned in the docs anywhere, so imo it isn't a fully supported feature and it's fine to remove it. 

Would also appreciate @patrickvonplaten's comment.

**Comment by ydshieh** [comment]
There are also `FunnelForPreTraining` and `ElectraForPreTraining` which use attention_mask

```
            if attention_mask is not None:
                active_loss = attention_mask.view(-1, discriminator_sequence_output.shape[1]) == 1
                active_logits = logits.view(-1, discriminator_sequence_output.shape[1])[active_loss]
                active_labels = labels[active_loss]
                loss = loss_fct(active_logits, active_labels.float())
```
(I saw it during the search for the last commit)

**Comment by ydshieh** [comment]
Before merging, would it be better to add a test? There is a more general test in #15256 (not merged yet, see below), which will fail (due to other PT/TF differences).

I can add that test in this PR, but restrict to `...ForTokenClassification` models here. The test can be made more general along the way of fixing other PT/TF differences.

https://github.com/ydshieh/transformers/blob/47a5d26b58afbfae2366328fe4242de9bb662fcb/tests/test_modeling_tf_common.py#L411-L461

**Comment by ydshieh** [comment]
Well, `FunnelForPreTraining` and `ElectraForPreTraining` use binary classification for pretraining objective.

```
            - 0 indicates the token is an original token,
            - 1 indicates the token was replaced.
```
and `nn.BCEWithLogitsLoss` doesn't support `ignore_index`.

The best one can do it probably just to make `TFFunnelForPreTraining` and `TFElectraForPreTraining` return loss (currently, they don't compute loss)

**Comment by LysandreJik** [comment]
Also cc @patil-suraj

File: src/transformers/models/albert/modeling_albert.py [hunk 1]
```diff
@@ -1150,16 +1150,7 @@ def forward(
         loss = None
         if labels is not None:
             loss_fct = CrossEntropyLoss()
-            # Only keep active parts of the loss
```
Comment (patrickvonplaten) [inline src/transformers/models/albert/modeling_albert.py:1153]: Instead of breaking it, could we maybe deprecate it? IMO, there are still quite some people that train models without using our `Trainer`, our `DataCollators`, or official examples. In such cases, I can see people passing labels to the tokenizer:

```
tokenizer(label_str, return_tensors="pt", padding=True)
```

=> in this case the labels would be padded with `0's` which if we remove this part won't be ignored anymore. So IMO it would be nice to go through a deprecation cycle here

Comment (ydshieh) [reply src/transformers/models/albert/modeling_albert.py:1153]: Sure, I will add deprecation.

Comment (sgugger) [reply src/transformers/models/albert/modeling_albert.py:1153]: I don't understand what you're saying @patrickvonplaten . For token classification tasks, labels can't go through the tokenizer like this without making any sense to the model.

Comment (sgugger) [reply src/transformers/models/albert/modeling_albert.py:1153]: Like this would encode labels `["O", "B-PER", "I-PER"]` in something completely different than the expected `[0, 1, 2]` and putting label indices as zeros mean the "O" class, not ignore.

Comment (ydshieh) [reply src/transformers/models/albert/modeling_albert.py:1153]: (I guess I somehow mixed the general token classification tasks with the language modeling task here ...)

Comment (ydshieh) [reply src/transformers/models/albert/modeling_albert.py:1153]: I also have a hard time to add deprecation - adding this means we want to allow PT models to use attention_mask (for now), but the enhanced test (pt/tf equivalence) will fail, because TF models don't use `attention_mask` (and the labels used for testing don't have `-100` - unless we change `def _prepare_for_class` in `test_modeling_tf_common.py`)

**Comment by ydshieh** [comment]
I think what @sgugger said makes sense,  https://github.com/huggingface/transformers/pull/15294#discussion_r794484953, and considering https://github.com/huggingface/transformers/pull/15294#discussion_r794507876, I think I won't have any action for now until some more feedback from HF side (once you get sometime :-) )

File: src/transformers/models/albert/modeling_albert.py (hunk 1 above)
Comment (patrickvonplaten) [reply src/transformers/models/albert/modeling_albert.py:1153]: Just took a look into the `run_ner.py`. Yes sorry, I didn't thought carefully enough - the use case I'm referring only really applies to the LM case.

Okey good for me then!

File: examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py [hunk 2]
```diff
@@ -902,16 +902,7 @@ def forward(
...
-            if attention_mask is not None:
-                active_loss = attention_mask.view(-1) == 1
-                active_logits = logits.view(-1, self.num_labels)
-                active_labels = torch.where(
-                    active_loss, labels.view(-1), torch.tensor(loss_fct.ignore_index).type_as(labels)
-                )
-                loss = loss_fct(active_logits, active_labels)
-            else:
-                loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
+            loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))
```
Comment (sgugger) [inline examples/research_projects/movement-pruning/emmental/modeling_bert_masked.py:905]: We don't touch research projects, so you should revert this change.

**Review by sgugger — APPROVED** [review]
We're good to go then, just revert the change to the research project please, as we let their contributors maintain them, and we can merge :-)

**Comment by ydshieh** [comment]
Reverted the change to research projects / Tests good / Ready to go :-) / Thanks!

**Comment by sgugger** [comment]
Thank *you* for your contribution :-)
